capsules/extra/src/bus.rs

```diff
@@ -149,25 +149,69 @@ impl<'a, S: SpiMasterDevice<'a>> SpiMasterBus<'a, S> {
 }
 
 impl<'a, S: SpiMasterDevice<'a>> Bus<'a> for SpiMasterBus<'a, S> {
-    fn set_addr(&self, addr_width: BusWidth, addr: usize) -> Result<(), ErrorCode> {
-        match addr_width {
-            BusWidth::Bits8 => self
-                .addr_buffer
-                .take()
-                .map_or(Err(ErrorCode::NOMEM), |buffer| {
+    fn set_addr(&self, addr_width: BusWidth, addr: u64) -> Result<(), ErrorCode> {
```

Comment:
Without looking too much at the general interface here, it seems like what we'd really want to have is the bus address be encapsulated within the enum itself, like so:

```
    fn set_addr(&self, addr: BusAddr) -> Result<(), ErrorCode> {
```

and

```
enum BusAddr {
    Bits8([u8; 1]),
    Bits16BE([u8; 2]),
    Bits16LE([u8; 2]),
}
```

@bradjc's suggestion about making it generic also makes sense to me. One implementation of the `BusAddr` trait could still have multiple variants at runtime, but when you know that all your connected peripherals have only one address width, you'd simply plug in the particular type that describes this address.

Something like this could work:

```
struct BusAddr8([u8; 1]);
impl BusAddr for BusAddr8 { ... }
struct BusAddr16BE([u8; 2]);
impl BusAddr for BusAddr16BE { ... }

enum MultiBusAddr {
    Bits8(BusAddr8),
    Bits16BE(BusAddr16BE),
    ...,
}
impl BusAddr for MultiBusAddr { ... }

impl<'a, S: SpiMasterDevice<'a>, A: BusAddr> Bus<'a, A> for SpiMasterBus<'a, S> {
    fn set_addr(&self, addr: impl Into<A>) -> Result<(), ErrorCode> {
```

Pretty sure that the types would need some more massaging to actually work in practice, just some ideas.

Reply:
I agree with enum, I'm not sure about the templating. Does it really make sense? What about adding a `write_to_slice` function to the enum?

Reply:
> I agree with enum, I'm not sure about the templating. Does it really make sense? What about adding a `write_to_slice` function to the enum?

That will still have the `enum BusAddr` be at least 9 bytes in size everywhere, which seems slightly wasteful (although not more wasteful than passing the `enum BusWidth` plus a `u64`).

I think this is a good step in the right direction, and then we can always change it to be a generic later.

Reply:
Oh I see what you mean now.

Reply:
I am wondering whether to use a `dyn Trait` instead of a generic? The bus might be used several times on a chip with different address sizes. The generic will duplicate code, while the `dye` trait will perform some small overhead. Not sure which one is better.

Reply:
Thanks for feedback! I will try to sketch out an implementation in the following days. I see this two possible implementations:

```rust
struct BusAddr<E: Endianess, const N: usize>([u8; N], PhantomData<E>);

trait BusAddrBytes<const N: usize> {
    fn bytes(&self) -> &[u8; N];
}
impl From<u16> for BusAddr<LittleEndian, 2> {
    fn from(value: u16) -> Self {
        Self([value as u8, (value >> 8) as u8], PhantomData)
    }
}
```

or 

```rust
struct BusAddr8([u8; 1]);
struct BusAddr16BE([u8; 2]);
trait BusAddrBytes {
    fn bytes(&self) -> &[u8];
}
impl From<u16> for BusAddr16LE {
    fn from(value: u16) -> Self {
        Self([value as u8, (value >> 8) as u8])
    }
}
...

```
Which one do you think works better?

Reply:
I like the second one better, as it explicitly pre-defines the possible address widths. I presume we don't want to support 3-byte addresses?

Also, I'm not sure about `From<u16>` (or `From`-based conversions from integers generally). I more so through of `From` to be useful for if we have a type that reconciles multiple addresses widths at runtime, but maybe that's too complex.

By the way, Rust offers `{to,from}_{le,be}_bytes` functions on integer types which take or produce an array directly. We should prefer those over bitshifts.

Reply:
I tried the new implementation. I'm waiting for feedback.